egs/audioset/AT/zipformer/train.py

```diff
@@ -0,0 +1,1189 @@
+        filename=filename,
+        model=model,
+        model_avg=model_avg,
+        params=params,
+        optimizer=optimizer,
+        scheduler=scheduler,
+        sampler=sampler,
+        scaler=scaler,
+        rank=rank,
+    )
+
+    if params.best_train_epoch == params.cur_epoch:
+        best_train_filename = params.exp_dir / "best-train-loss.pt"
+        copyfile(src=filename, dst=best_train_filename)
+
+    if params.best_valid_epoch == params.cur_epoch:
+        best_valid_filename = params.exp_dir / "best-valid-loss.pt"
+        copyfile(src=filename, dst=best_valid_filename)
+
+
+def compute_loss(
+    params: AttributeDict,
+    model: Union[nn.Module, DDP],
+    batch: dict,
+    is_training: bool,
+) -> Tuple[Tensor, MetricsTracker]:
+    """
+    Compute loss given the model and its inputs.
+
+    Args:
+      params:
+        Parameters for training. See :func:`get_params`.
+      model:
+        The model for training. It is an instance of Zipformer in our case.
+      batch:
+        A batch of data. See `lhotse.dataset.K2SpeechRecognitionDataset()`
+        for the content in it.
+      is_training:
+        True for training. False for validation. When it is True, this
+        function enables autograd during computation; when it is False, it
+        disables autograd.
+     warmup: a floating point value which increases throughout training;
+        values >= 1.0 are fully warmed up and have all modules present.
+    """
+    device = model.device if isinstance(model, DDP) else next(model.parameters()).device
+    feature = batch["inputs"]
+    # at entry, feature is (N, T, C)
+    assert feature.ndim == 3
+    feature = feature.to(device)
+
+    supervisions = batch["supervisions"]
+    events = supervisions["audio_event"]  # the label indices are in CED format
+    labels, _ = str2multihot(events, n_classes=params.num_events)
+    labels = labels.to(device)
+
+    feature_lens = supervisions["num_frames"].to(device)
+
+    batch_idx_train = params.batch_idx_train
+    warm_step = params.warm_step
+
+    with torch.set_grad_enabled(is_training):
+        audio_tagging_loss = model(
+            x=feature,
+            x_lens=feature_lens,
+            target=labels,
+        )
+
+        loss = 0.0
+        loss += audio_tagging_loss
+
+    assert loss.requires_grad == is_training
+
+    info = MetricsTracker()
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        info["frames"] = (feature_lens // params.subsampling_factor).sum().item()
+
+    # Note: We use reduction=sum while computing the loss.
+    info["loss"] = loss.detach().cpu().item()
+    info["audio_tagging_loss"] = audio_tagging_loss.detach().cpu().item()
+
+    return loss, info
+
+
+def str2multihot(events: List[str], n_classes=527, id_mapping=None):
+    # Convert strings separated by semi-colon to multi-hot class labels
+    # input: ["1;2", "2;3"]
+    # output: torch.tensor([[1,1,0], [0,1,1]])
```

Comment:
Could you specify the `n_classes` in this example?

The output does not look correct to me.

Reply:
This dummy example is for `n_classes=3`, the class ID should start from 1.

Reply:
>  the class ID should start from 1.

Is it also starting from 1 for the training data?

If not,  I suggest that you use a consistent starting value. Otherwise, it is confusing.